enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,256 @@
+
+The `bootstrap-in-place.ign` will be embedded into an RHCOS liveCD using coreos-install embed command.
+
+The user will boot a machine with this liveCD, the liveCD will start executing a similar flow as a bootstrap node in a regular installation:
+bootkube.sh will do all the rendering logic.
+The bootstrap static pods should be generated in a way that the control plane operators will be able to identify them and either continues in a controlled way for the next revision, or just keep them as the correct revision and reuse them.
+
+cluster-bootstrap will apply all the required manifests (under /opt/openshift/manifests/)
+
+Bootkube will get the master ignition from machine-config-server and
+enrich the master ignition with control plane static pods manifests and all required resources including etcd data.
+Bootkube.sh will write the master ignition along with rhcos to disk.
+At this point bootkube will reboot the node and let it finish the cluster creation.
+
+Post reboot:
+The kubelet service will start the control plane static pods.
+Kubelet will send a CSR (see below) and join the cluster.
+CVO will deploy all cluster operators.
+The control plane operators will rollout a new revision (if necessary).  
+
+#### Openshift-installer
+
+We will add logic to the installer to create `bootstrap-in-place.ign` ignition config.
+This ignition config will diverge from the bootstrap ignitionin:
+bootkube:
+1. Start cluster-bootstrap without required pods (`--required-pods=''`)
+2. Run cluster-bootstrap with iBIP entrypoint to enrich the master ignition.
+3. Write RHCOS image and the master ignition to disk.
+4. Reboot the node.
+
+#### Cluster-bootstrap
+
+By default, `cluster-bootstrap` start the bootstrap control plane and create all the manifests under /opt/openshift/manifests.
+`cluster-bootstrap` also wait for a list required pods to be ready, these pods are expetced to start running on the master nodes.
+In case we are running the bootstrap in place, there is no master node that can run the pods. cluster-bootstrap should apply the manifest and tear down the control plane, in case it fail to apply some of the manifest it should return an error.
+
+
+`cluster-bootstrap` will have a new entrypoint `iBIP` which will get the master ignition as input and will enrich the master ignition with control plane static pods manifests and all required resources including etcd data.
+
+#### Bootstrap / Control plane static pods
+The control plane components we will add to the master ignition are (to be placed under `/etc/kubernetes/manifests`):
+1. etcd-pod
+2. kube-apiserver-pod
+3. kube-controller-manager-pod
+4. kube-scheduler-pod
+
+Control plane required resources to be added to the ignition:
+1. `/var/lib/etcd`
+2. `/etc/kubernetes/bootstrap-configs`
+3. /opt/openshift/tls/* (`/etc/kubernetes/bootstrap-secrets`)
+4. /opt/openshift/auth/kubeconfig-loopback (`/etc/kubernetes/bootstrap-secrets/kubeconfig`)
```

Comment:
Who starts the cluster-version operator up post-reboot?

Reply:
Yes, there is a CVO deployment in etcd.
The above mentioned static pods get replaced by the static pods deployed by the operators